pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/typedefinition/JavaTypeDefinition.java

```diff
@@ -106,254 +84,80 @@ public static int getGenericTypeIndex(TypeVariable<?>[] typeParameters, final St
-        // proper vs proper
+    public abstract Set<JavaTypeDefinition> getSuperTypeSet();
 
-        // Note: we have to force raw types to compute their generic args, class Stuff<? extends List<Stuff>>
-        // Stuff a;
-        // Stuff<? extends List<Stuff>> b;
-        // Stuff<List<Stuff>> c;
-        // all of the above should be equal
+    protected abstract Set<JavaTypeDefinition> getSuperTypeSet(Set<JavaTypeDefinition> destinationSet);
 
-        for (int i = 0; i < getTypeParameterCount(); ++i) {
-            // Note: we assume that cycles can only exist because of raw types
-            if (!getGenericType(i).equals(otherTypeDef.getGenericType(i))) {
-                return false;
-            }
-        }
+    public abstract Set<Class<?>> getErasedSuperTypeSet();
 
-        return true;
+    /**
+     * @return true if clazz is generic and had not been parameterized
+     */
+    public final boolean isRawType() {
+        return !isIntersectionType() && isGeneric() && CLASS_EXACT_TYPE_DEF_CACHE.containsKey(getType());
     }
 
-    @Override
-    public int hashCode() {
-        return clazz.hashCode();
-    }
+    public abstract JavaTypeDefinition getAsSuper(Class<?> superClazz);
 
-    public Set<JavaTypeDefinition> getSuperTypeSet() {
-        return getSuperTypeSet(new HashSet<JavaTypeDefinition>());
+    public final boolean isExactType() {
+        return definitionType == TypeDefinitionType.EXACT;
     }
 
-    private Set<JavaTypeDefinition> getSuperTypeSet(Set<JavaTypeDefinition> destinationSet) {
-        destinationSet.add(this);
-
-        if (this.clazz != Object.class) {
-
-            resolveTypeDefinition(clazz.getGenericSuperclass()).getSuperTypeSet(destinationSet);
-
-            for (Type type : clazz.getGenericInterfaces()) {
-                resolveTypeDefinition(type).getSuperTypeSet(destinationSet);
-            }
-        }
-
-        return destinationSet;
+    public final boolean isUpperBound() {
+        return definitionType == TypeDefinitionType.UPPER_BOUND
+                || definitionType == TypeDefinitionType.UPPER_WILDCARD;
     }
 
-    public Set<Class<?>> getErasedSuperTypeSet() {
-        Set<Class<?>> result = new HashSet<>();
-        result.add(Object.class);
-        return getErasedSuperTypeSet(this.clazz, result);
+    public final boolean isLowerBound() {
+        return definitionType == TypeDefinitionType.LOWER_WILDCARD;
     }
 
-    private static Set<Class<?>> getErasedSuperTypeSet(Class<?> clazz, Set<Class<?>> destinationSet) {
-        if (clazz != null) {
-            destinationSet.add(clazz);
-            getErasedSuperTypeSet(clazz.getSuperclass(), destinationSet);
-
-            for (Class<?> interfaceType : clazz.getInterfaces()) {
-                getErasedSuperTypeSet(interfaceType, destinationSet);
-            }
-        }
-
-        return destinationSet;
+    public final boolean isIntersectionType() {
+        return getJavaTypeCount() != 1;
```

Comment:
Intersection type means the intersection of two or more types. So like Integer & Runnable. The reason it can return false, is because ? super Something, T extends Stuff, T extends Stuff & Other etc. These are all represented by a JavaTypeDefinitionIntersection object containing JavaTypeDefinitionSimple objects (1 or more). It is like this so as to avoid making the type hierarchy more complex.

But you are right in that the name is miss leading. Can't think of something better...